server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -983,8 +987,25 @@ private Engine.IndexResult applyIndexOperation(
             verifyNotClosed(e);
             return new Engine.IndexResult(e, version, opPrimaryTerm, seqNo, sourceToParse.id());
         }
+        Engine.IndexResult indexResult = index(engine, operation);
+        if (indexResult.isCreated()) {
+            updateFieldHasValueWithFieldsFromParsedDoc(operation.parsedDoc());
+        }
+        return indexResult;
```

Comment:
I think that I am getting confused by the diff here: why the additional index call? Intuitively, we would only need to update the set for each doc?

Reply:
before it was: `return index(engine, operation);` therefore we call index() and then return the result, in my change I am calling index, storing the result, updating the `fieldHasValue` set if the operation is successful and returning the result. This has being done in order to make sure to add fields to the set iff the index operation is successful. (Yep, git diff is not the best here)